### ISSUE TITLE
Add type annotations to public interfaces

### ISSUE THREAD

**Issue by numberoverzero** [numberoverzero/bloop#50]
This will require dropping 3.4 support, but I think that's reasonable.  There are very few considerations when [Porting to Python 3.5](https://docs.python.org/3/whatsnew/3.5.html#porting-to-python-3-5) and in practice, I've never seen an issue through a dozen or so migrations.

Bloop will need to consider a policy in the future for minimum required python version.  For the foreseeable future, there aren't any compelling features for bloop specifically in 3.6 or 3.7.  A rough draft would ensure that minimum python version bumps would correspond to a major version bump (ie. 3.6 could be required for bloop 2.0.0).

The following need type annotations:

- [ ] `SessionWrapper`
- [ ] `Engine`
- [ ] `Column`
- [ ] `Condition`
- [ ] `Type`
- [ ] `Index`
- [ ] `BaseModel`
- [ ] `Scan`, `Query`, `Search`
- [ ] `ScanIterator`, `QueryIterator`, `SearchIterator`
- [ ] `PreparedSearch`

Internal methods should also be annotated; due to the dynamic nature of many methods, I expect non-public methods to use a lot of `Union`, `Optional`, and `Any`.

**Comment by numberoverzero** [comment]
tl;dr This may not happen.  The typing module does not make this easy, and I'm not sure the pep even knows what it wants.


After a naive attempt at this, it turns out circular imports are painful.  I think I can do typing with strings, but that kind of sucks.  I'll review what other packages are doing, but the notes in pep 484 [don't inspire confidence](https://www.python.org/dev/peps/pep-0484/#the-type-of-class-objects); it seems like this is not as solid a feature as I first expected, but an evolving spec that hasn't finalized at all.

From that document, I **think** I could use `ColumnType = TypeVar("ColumnType", bound="Column")` to specify all of the types without any imports.  But, it's still not clear if I can do that with strings.  If I can use strings for the bound param, it's not clear how I'll hook those `TypeVar` up to the class definitions without importing into the definition (again, that would make the imports circular).

**Comment by numberoverzero** [comment]
Closing for now.  It may worthwhile at some point in the future, but there are a lot of cyclic references that would either require more imports, or defining the types in a different file and importing them everywhere.  That seems less than useful.
